File: public/data_enrichment/semantic_search.py
import json
import os
import psycopg2
from psycopg2.extras import RealDictCursor
import google.generativeai as genai
import sys
import re
from typing import List, Dict, Optional
from flask import Flask, request, jsonify
from flask_cors import CORS

# Load configuration
def load_config():
    """Load configuration from config.json"""
    config_path = os.path.join(os.path.dirname(__file__), 'config.json')
    try:
        with open(config_path, 'r') as f:
            config = json.load(f)
        return config
    except FileNotFoundError:
        print(f"Error: config.json not found at {config_path}")
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in config.json: {e}")
        sys.exit(1)

# Load config
CONFIG = load_config()
DATABASE_URL = CONFIG['database']['url']

# Gemini Configuration
GEMINI_API_KEY = os.getenv("GEMINI_API_KEY", CONFIG['gemini']['api_key'])
EMBEDDING_MODEL = CONFIG['gemini'].get('embedding_model', 'models/embedding-001')
EMBEDDING_DIMENSIONS = CONFIG['gemini'].get('embedding_dimensions', 768)

# Configure Gemini
if not GEMINI_API_KEY or GEMINI_API_KEY == "YOUR_GEMINI_API_KEY_HERE":
    print("Error: GEMINI_API_KEY not set. Please set it in config.json or as environment variable.")
    sys.exit(1)

genai.configure(api_key=GEMINI_API_KEY)
print(f"Using Gemini embedding model: {EMBEDDING_MODEL} ({EMBEDDING_DIMENSIONS} dimensions)")

# Flask app
app = Flask(__name__)
CORS(app)  # Enable CORS for frontend access

def preprocess_query(text: str) -> str:
    """
    Preprocess query text to match how embeddings were created.
    Replace underscores with spaces to match embedding format.
    Example: "water_proof" -> "water proof"
    """
    if not text or not text.strip():
        return ""
    
    # Replace underscores with spaces (to match embedding format)
    text = text.replace('_', ' ')
    
    # Normalize whitespace
    text = re.sub(r'\s+', ' ', text)
    
    return text.strip()

def format_embedding_for_db(embedding: List[float]) -> str:
    """Convert embedding list to PostgreSQL vector format"""
    if not embedding:
        return None
    # Format: [0.1,0.2,0.3,...] for pgvector
    return '[' + ','.join(str(float(x)) for x in embedding) + ']'

def get_embedding(text: str, task_type: str = "RETRIEVAL_QUERY") -> Optional[List[float]]:
    """Generate embedding for query text using Gemini"""
    if not text or not text.strip():
        return None
    
    # Preprocess query to match embedding format
    processed_text = preprocess_query(text)
    
    if not processed_text:
        return None
    
    try:
        # Use Gemini embedding model
        # RETRIEVAL_QUERY for search queries, RETRIEVAL_DOCUMENT for documents
        result = genai.embed_content(
            model=EMBEDDING_MODEL,
            content=processed_text,
            task_type=task_type
        )
        embedding = result['embedding']
        
        # Validate dimensions
        if embedding and len(embedding) == EMBEDDING_DIMENSIONS:
            return embedding
        else:
            print(f"Warning: Embedding has {len(embedding) if embedding else 0} dimensions, expected {EMBEDDING_DIMENSIONS}")
            return embedding
    except Exception as e:
        print(f"Error generating embedding: {e}")
        return None

def semantic_search_structured(query_fields: Dict[str, str], limit: int = 10, min_score: float = 0.6) -> List[Dict]:
    """
    Perform structured semantic search with field-specific matching.
    
    Args:
        query_fields: Dictionary with optional fields:
            - use_case: Query for use_cases embeddings (highest priority)
            - best_for: Query for best_for embeddings (medium priority)
            - not_best_for: Query for not_best_for embeddings (low priority)
            - tradeoff: Query for tradeoffs embeddings (low priority)
            - ambiguity_triggers: Query for ambiguity_triggers embeddings (low priority)
        limit: Maximum number of results to return
        min_score: Minimum similarity score threshold
    
    Priority weights:
        - use_case: 1.0 (highest)
        - best_for: 0.7 (medium)
        - not_best_for, tradeoff, ambiguity_triggers: 0.3 (lowest)
    """
    
    # Priority weights for each field type
    PRIORITY_WEIGHTS = {
        'use_cases': 1.0,        # Highest priority
        'best_for': 0.7,         # Medium priority
        'not_best_for': 0.3,     # Low priority
        'tradeoffs': 0.3,        # Low priority
        'ambiguity_triggers': 0.3 # Low priority
    }
    
    # Field mapping: user field name -> database embedding_type
    FIELD_MAPPING = {
        'use_case': 'use_cases',
        'best_for': 'best_for',
        'not_best_for': 'not_best_for',
        'tradeoff': 'tradeoffs',
        'ambiguity_triggers': 'ambiguity_triggers'
    }
    
    # Generate embeddings for all provided query fields
    query_embeddings = {}
    for field_name, query_text in query_fields.items():
        if query_text and query_text.strip():
            embedding = get_embedding(query_text.strip())
            if embedding:
                db_field_name = FIELD_MAPPING.get(field_name, field_name)
                query_embeddings[db_field_name] = {
                    'embedding': embedding,
                    'embedding_str': format_embedding_for_db(embedding),
                    'weight': PRIORITY_WEIGHTS.get(db_field_name, 0.5),
                    'query_text': query_text.strip()
                }
    
    if not query_embeddings:
        return []
    
    conn = psycopg2.connect(DATABASE_URL)
    cursor = conn.cursor(cursor_factory=RealDictCursor)
    
    # Store matches per variant per field type
    variant_matches = {}  # {variant_id: {field_type: {best_match_info}}}
    
    try:
        # Search each field type separately and collect matches
        for field_type, query_info in query_embeddings.items():
            embedding_str = query_info['embedding_str']
            weight = query_info['weight']
            
            # Search for matches in this specific field type
            cursor.execute("""
                SELECT 
                    ke.variant_id,
                    ke.product_id,
                    ke.embedding_type,
                    ke.embedding_text,
                    1.0 * (1 - (ke.embedding <=> %s::vector)) as similarity_score
                FROM public.kb_embeddings ke
                WHERE ke.embedding_type = %s
                AND (1 - (ke.embedding <=> %s::vector)) >= %s
                ORDER BY ke.embedding <=> %s::vector
                LIMIT %s
            """, (embedding_str, field_type, embedding_str, min_score, embedding_str, limit * 2))
            
            field_results = cursor.fetchall()
            
            # Store best match per variant for this field type
            for result in field_results:
                variant_id = result['variant_id']
                similarity = float(result['similarity_score'])
                
                # Calculate weighted score for this match
                weighted_score = similarity * weight
                
                if variant_id not in variant_matches:
                    variant_matches[variant_id] = {}
                
                # Store the best match for this field type (highest weighted score)
                if field_type not in variant_matches[variant_id] or \
                   weighted_score > variant_matches[variant_id][field_type]['weighted_score']:
                    variant_matches[variant_id][field_type] = {
                        'embedding_text': result['embedding_text'],
                        'similarity_score': similarity,
                        'weighted_score': weighted_score,
                        'weight': weight
                    }
        
        if not variant_matches:
            return []
        
        # Get variant IDs that have at least one match
        variant_ids = list(variant_matches.keys())
        
        # Get full variant details including confidence_score
        placeholders = ','.join(['%s'] * len(variant_ids))
        cursor.execute(f"""
            SELECT 
                kev.variant_id,
                kev.product_id,
                kev.use_cases,
                kev.skill_level,
                kev.portability_score,
                kev.price_tier,
                kev.best_for,
                kev.not_best_for,
                kev.tradeoffs,
                kev.confidence_score,
                COALESCE(kev.confidence_score, 50) as normalized_confidence,
                lp.title as product_title,
                lp.handle as product_handle
            FROM public.kb_enriched_variants kev
            LEFT JOIN public.latest_product lp ON kev.product_id = lp.id
            WHERE kev.variant_id IN ({placeholders})
        """, variant_ids)
        
        variant_details = {r['variant_id']: dict(r) for r in cursor.fetchall()}
        
        # Calculate final scores for each variant
        results = []
        
        for variant_id, field_matches in variant_matches.items():
            if variant_id not in variant_details:
                continue
            
            # Calculate aggregate score
            total_weighted_score = 0.0
            total_weight = 0.0
            match_count = 0
            
            # Collect match information
            match_details = {}
            
            for field_type, match_info in field_matches.items():
                total_weighted_score += match_info['weighted_score']
                total_weight += match_info['weight']
                match_count += 1
                match_details[field_type] = {
                    'embedding_text': match_info['embedding_text'],
                    'similarity_score': match_info['similarity_score']
                }
            
            # Average weighted score (normalized by total weights)
            if total_weight > 0:
                final_score = total_weighted_score / total_weight
            else:
                final_score = 0.0
            
            # Boost score based on number of matches (variants matching multiple criteria are better)
            match_bonus = 0.1 * min(match_count / len(query_embeddings), 1.0)
            final_score = min(final_score + match_bonus, 1.0)
            
            # Add confidence score influence (10% weight)
            confidence = variant_details[variant_id].get('normalized_confidence', 50)
            confidence_factor = (confidence / 100.0) * 0.1
            final_score = min(final_score + confidence_factor, 1.0)
            
            # Build result dictionary
            result_dict = {
                'variant_id': variant_id,
                'product_id': variant_details[variant_id]['product_id'],
                'final_score': final_score,
                'match_count': match_count,
                'total_fields_searched': len(query_embeddings),
                'matches': match_details,
                'confidence_score': variant_details[variant_id].get('confidence_score')
            }
            result_dict.update(variant_details[variant_id])
            results.append(result_dict)
        
        # Sort by final score (highest first)
        results.sort(key=lambda x: -x['final_score'])
        
        # Return top results
        return results[:limit]
        
    finally:
        cursor.close()
        conn.close()

@app.route('/search', methods=['POST'])
def search_endpoint():
    """
    API endpoint for structured semantic search
    
    Accepts structured query with optional fields:
    {
        "use_case": "under water recording",
        "best_for": "water sports players",
        "not_best_for": "deep sea diving below 100m",
        "tradeoff": "4k video is capped at 30fps",
        "ambiguity_triggers": "specific sensor or processor model is missing",
        "limit": 10,
        "min_score": 0.6
    }
    
    All fields except limit and min_score are optional.
    """
    try:
        data = request.get_json()
        
        if not data:
            return jsonify({'error': 'Request body is required'}), 400
        
        # Extract query fields (all optional)
        query_fields = {}
        if data.get('use_case'):
            query_fields['use_case'] = str(data['use_case']).strip()
        if data.get('best_for'):
            query_fields['best_for'] = str(data['best_for']).strip()
        if data.get('not_best_for'):
            query_fields['not_best_for'] = str(data['not_best_for']).strip()
        if data.get('tradeoff'):
            query_fields['tradeoff'] = str(data['tradeoff']).strip()
        if data.get('ambiguity_triggers'):
            query_fields['ambiguity_triggers'] = str(data['ambiguity_triggers']).strip()
        
        # At least one query field must be provided
        if not query_fields:
            return jsonify({
                'error': 'At least one query field is required. Provide: use_case, best_for, not_best_for, tradeoff, or ambiguity_triggers'
            }), 400
        
        limit = int(data.get('limit', 10))
        min_score = float(data.get('min_score', 0.6))
        
        results = semantic_search_structured(query_fields, limit=limit, min_score=min_score)
        
        return jsonify({
            'query_fields': query_fields,
            'count': len(results),
            'min_score': min_score,
            'results': results
        })
        
    except ValueError as e:
        return jsonify({'error': f'Invalid parameter: {str(e)}'}), 400
    except Exception as e:
        import traceback
        traceback.print_exc()
        return jsonify({'error': str(e)}), 500

@app.route('/', methods=['GET'])
def root():
    """Root endpoint with API information"""
    return jsonify({
        'service': 'Semantic Search API',
        'version': '1.0',
        'endpoints': {
            'search': {
                'path': '/search',
                'method': 'POST',
                'description': 'Perform structured semantic search on product variants',
                'body': {
                    'use_case': 'string (optional) - Query for use_cases embeddings (highest priority)',
                    'best_for': 'string (optional) - Query for best_for embeddings (medium priority)',
                    'not_best_for': 'string (optional) - Query for not_best_for embeddings (low priority)',
                    'tradeoff': 'string (optional) - Query for tradeoffs embeddings (low priority)',
                    'ambiguity_triggers': 'string (optional) - Query for ambiguity_triggers embeddings (low priority)',
                    'limit': 'integer (optional, default: 10) - Number of results',
                    'min_score': 'float (optional, default: 0.6) - Minimum similarity score',
                    'note': 'At least one query field (use_case, best_for, etc.) is required'
                },
                'example': {
                    'use_case': 'under water recording',
                    'best_for': 'water sports players',
                    'not_best_for': 'deep sea diving below 100m',
                    'limit': 10
                }
            },
            'health': {
                'path': '/health',
                'method': 'GET',
                'description': 'Health check endpoint'
            }
        },
        'model': 'gemini',
        'embedding_model': EMBEDDING_MODEL,
        'embedding_dimensions': EMBEDDING_DIMENSIONS
    })

@app.route('/health', methods=['GET'])
def health_check():
    """Health check endpoint"""
    return jsonify({'status': 'healthy', 'model': 'gemini', 'embedding_model': EMBEDDING_MODEL})

if __name__ == '__main__':
    port = int(os.getenv('PORT', 5500))
    print(f"Starting semantic search service on port {port}...")
    print(f"Using Gemini embedding model: {EMBEDDING_MODEL}")
    app.run(host='0.0.0.0', port=port, debug=True)

File: lib/semanticSearch.ts
/**
 * Semantic Search Module
 *
 * Port of semantic_search.py to TypeScript
 * Provides priority-weighted semantic search over kb_embeddings and kb_enriched_variants
 */

import { Pool, PoolClient } from "pg";
import { generateEmbedding } from "./embeddings";

// ============================================================================
// PRIORITY WEIGHTS
// ============================================================================

const PRIORITY_WEIGHTS: Record<string, number> = {
  use_cases: 1.0, // Highest priority
  best_for: 0.7, // Medium priority
  not_best_for: 0.3, // Low priority
  tradeoffs: 0.3, // Low priority
  ambiguity_triggers: 0.3, // Low priority
};

// Field mapping: user field name -> database embedding_type
const FIELD_MAPPING: Record<string, string> = {
  use_case: "use_cases",
  best_for: "best_for",
  not_best_for: "not_best_for",
  tradeoff: "tradeoffs",
  ambiguity_triggers: "ambiguity_triggers",
};

// ============================================================================
// DATABASE CONNECTION
// ============================================================================

let pool: Pool | null = null;

function getPool(): Pool | null {
  const databaseUrl =
    process.env.DATABASE_URL || process.env.SUPABASE_DB_URL || "";

  if (!databaseUrl) {
    console.warn("[SemanticSearch] DATABASE_URL not set.");
    return null;
  }

  if (!pool) {
    pool = new Pool({
      connectionString: databaseUrl,
      ssl:
        process.env.DB_SSL === "false"
          ? false
          : { rejectUnauthorized: false },
    });
  }
  return pool;
}

async function getClient(): Promise<PoolClient | null> {
  const p = getPool();
  if (!p) return null;
  return p.connect();
}

// ============================================================================
// TYPES
// ============================================================================

export interface SemanticSearchQuery {
  use_case?: string;
  best_for?: string;
  not_best_for?: string;
  tradeoff?: string;
  ambiguity_triggers?: string;
}

export interface MatchDetails {
  embedding_text: string;
  similarity_score: number;
}

export interface EnrichedVariant {
  variant_id: string;
  product_id: string;
  use_cases: string[];
  skill_level: string | null;
  portability_score: number | null;
  price_tier: string | null;
  best_for: string[];
  not_best_for: string[];
  tradeoffs: string[];
  ambiguity_triggers: string[];
  confidence_score: number | null;
  // Product details from latest_product
  product_title: string;
  product_handle: string | null;
  vendor: string | null;
  product_type: string | null;
  body_html: string | null;
  images: unknown | null;
  variants: unknown | null;
}

export interface SemanticSearchResult extends EnrichedVariant {
  final_score: number;
  match_count: number;
  total_fields_searched: number;
  matches: Record<string, MatchDetails>;
}

export interface SemanticSearchOptions {
  limit?: number;
  min_score?: number;
  skill_level?: string;
  price_tier?: string;
  budget_min?: number;
  budget_max?: number;
}

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

/**
 * Preprocess query text to match how embeddings were created.
 * Replace underscores with spaces to match embedding format.
 */
function preprocessQuery(text: string): string {
  if (!text || !text.trim()) return "";
  // Replace underscores with spaces (to match embedding format)
  let processed = text.replace(/_/g, " ");
  // Normalize whitespace
  processed = processed.replace(/\s+/g, " ");
  return processed.trim();
}

/**
 * Format embedding array to PostgreSQL vector format
 */
function formatEmbeddingForDb(embedding: number[]): string {
  if (!embedding || embedding.length === 0) return "";
  return "[" + embedding.map((x) => x.toString()).join(",") + "]";
}

// ============================================================================
// MAIN SEARCH FUNCTION
// ============================================================================

/**
 * Perform structured semantic search with field-specific matching.
 *
 * Priority weights:
 * - use_case: 1.0 (highest)
 * - best_for: 0.7 (medium)
 * - not_best_for, tradeoff, ambiguity_triggers: 0.3 (lowest)
 */
export async function semanticSearchStructured(
  queryFields: SemanticSearchQuery,
  options: SemanticSearchOptions = {}
): Promise<SemanticSearchResult[]> {
  const { limit = 10, min_score = 0.5, skill_level, price_tier, budget_min, budget_max } = options;

  // Generate embeddings for all provided query fields
  const queryEmbeddings: Record<
    string,
    {
      embedding: number[];
      embeddingStr: string;
      weight: number;
      queryText: string;
    }
  > = {};

  for (const [fieldName, queryText] of Object.entries(queryFields)) {
    if (queryText && queryText.trim()) {
      const processedText = preprocessQuery(queryText.trim());
      if (!processedText) continue;

      try {
        const embedding = await generateEmbedding(processedText);
        if (embedding && embedding.length > 0) {
          const dbFieldName = FIELD_MAPPING[fieldName] || fieldName;
          queryEmbeddings[dbFieldName] = {
            embedding,
            embeddingStr: formatEmbeddingForDb(embedding),
            weight: PRIORITY_WEIGHTS[dbFieldName] || 0.5,
            queryText: processedText,
          };
        }
      } catch (error) {
        console.error(
          `[SemanticSearch] Error generating embedding for ${fieldName}:`,
          error
        );
      }
    }
  }

  if (Object.keys(queryEmbeddings).length === 0) {
    console.warn("[SemanticSearch] No valid query embeddings generated");
    return [];
  }

  const client = await getClient();
  if (!client) {
    console.warn("[SemanticSearch] No database client available");
    return [];
  }

  try {
    // Store matches per variant per field type
    const variantMatches: Record<
      string,
      Record<
        string,
        {
          embedding_text: string;
          similarity_score: number;
          weighted_score: number;
          weight: number;
        }
      >
    > = {};

    // Search each field type separately and collect matches
    for (const [fieldType, queryInfo] of Object.entries(queryEmbeddings)) {
      const { embeddingStr, weight } = queryInfo;

      const result = await client.query(
        `
        SELECT 
          ke.variant_id::text as variant_id,
          ke.product_id::text as product_id,
          ke.embedding_type,
          ke.embedding_text,
          1.0 * (1 - (ke.embedding <=> $1::vector)) as similarity_score
        FROM public.kb_embeddings ke
        WHERE ke.embedding_type = $2
        AND (1 - (ke.embedding <=> $1::vector)) >= $3
        ORDER BY ke.embedding <=> $1::vector
        LIMIT $4
        `,
        [embeddingStr, fieldType, min_score, limit * 3]
      );

      // Store best match per variant for this field type
      for (const row of result.rows) {
        const variantId = row.variant_id;
        const similarity = parseFloat(row.similarity_score);
        const weightedScore = similarity * weight;

        if (!variantMatches[variantId]) {
          variantMatches[variantId] = {};
        }

        // Store the best match for this field type (highest weighted score)
        if (
          !variantMatches[variantId][fieldType] ||
          weightedScore > variantMatches[variantId][fieldType].weighted_score
        ) {
          variantMatches[variantId][fieldType] = {
            embedding_text: row.embedding_text,
            similarity_score: similarity,
            weighted_score: weightedScore,
            weight,
          };
        }
      }
    }

    if (Object.keys(variantMatches).length === 0) {
      console.log("[SemanticSearch] No matching variants found");
      return [];
    }

    // Get variant IDs that have at least one match
    const variantIds = Object.keys(variantMatches);

    // Build WHERE clause for filters
    const whereClauses: string[] = [`kev.variant_id = ANY($1::bigint[])`];
    const params: unknown[] = [variantIds.map((id) => BigInt(id))];
    let paramIndex = 2;

    if (skill_level) {
      whereClauses.push(`kev.skill_level = $${paramIndex}`);
      params.push(skill_level);
      paramIndex++;
    }

    if (price_tier) {
      whereClauses.push(`kev.price_tier = $${paramIndex}`);
      params.push(price_tier);
      paramIndex++;
    }

    // Get full variant details including product info
    const variantQuery = `
      SELECT 
        kev.variant_id::text as variant_id,
        kev.product_id::text as product_id,
        kev.use_cases,
        kev.skill_level,
        kev.portability_score,
        kev.price_tier,
        kev.best_for,
        kev.not_best_for,
        kev.tradeoffs,
        kev.ambiguity_triggers,
        COALESCE(kev.confidence_score, 50) as confidence_score,
        lp.title as product_title,
        lp.handle as product_handle,
        lp.vendor,
        lp.product_type,
        lp.body_html,
        lp.images,
        lp.variants
      FROM public.kb_enriched_variants kev
      LEFT JOIN public.latest_product lp ON kev.product_id = lp.id
      WHERE ${whereClauses.join(" AND ")}
    `;

    const variantResult = await client.query(variantQuery, params);

    const variantDetails: Record<string, EnrichedVariant> = {};
    for (const row of variantResult.rows) {
      variantDetails[row.variant_id] = {
        variant_id: row.variant_id,
        product_id: row.product_id,
        use_cases: row.use_cases || [],
        skill_level: row.skill_level,
        portability_score: row.portability_score,
        price_tier: row.price_tier,
        best_for: row.best_for || [],
        not_best_for: row.not_best_for || [],
        tradeoffs: row.tradeoffs || [],
        ambiguity_triggers: row.ambiguity_triggers || [],
        confidence_score: row.confidence_score,
        product_title: row.product_title || "Unknown Product",
        product_handle: row.product_handle,
        vendor: row.vendor,
        product_type: row.product_type,
        body_html: row.body_html,
        images: row.images,
        variants: row.variants,
      };
    }

    // Calculate final scores for each variant
    const results: SemanticSearchResult[] = [];

    for (const [variantId, fieldMatches] of Object.entries(variantMatches)) {
      if (!variantDetails[variantId]) {
        continue;
      }

      // Calculate aggregate score
      let totalWeightedScore = 0;
      let totalWeight = 0;
      let matchCount = 0;

      // Collect match information
      const matchDetails: Record<string, MatchDetails> = {};

      for (const [fieldType, matchInfo] of Object.entries(fieldMatches)) {
        totalWeightedScore += matchInfo.weighted_score;
        totalWeight += matchInfo.weight;
        matchCount++;
        matchDetails[fieldType] = {
          embedding_text: matchInfo.embedding_text,
          similarity_score: matchInfo.similarity_score,
        };
      }

      // Average weighted score (normalized by total weights)
      let finalScore = totalWeight > 0 ? totalWeightedScore / totalWeight : 0;

      // Boost score based on number of matches (variants matching multiple criteria are better)
      const matchBonus =
        0.1 * Math.min(matchCount / Object.keys(queryEmbeddings).length, 1.0);
      finalScore = Math.min(finalScore + matchBonus, 1.0);

      // Add confidence score influence (10% weight)
      const confidence = variantDetails[variantId].confidence_score || 50;
      const confidenceFactor = (confidence / 100.0) * 0.1;
      finalScore = Math.min(finalScore + confidenceFactor, 1.0);

      // Apply budget filter if provided
      if (budget_min !== undefined || budget_max !== undefined) {
        // Get price from variants JSON
        const variants = variantDetails[variantId].variants;
        if (variants && Array.isArray(variants) && variants.length > 0) {
          const price = parseFloat(variants[0]?.price || "0");
          if (budget_min !== undefined && price < budget_min) continue;
          if (budget_max !== undefined && price > budget_max) continue;
        }
      }

      results.push({
        ...variantDetails[variantId],
        final_score: finalScore,
        match_count: matchCount,
        total_fields_searched: Object.keys(queryEmbeddings).length,
        matches: matchDetails,
      });
    }

    // Sort by final score (highest first)
    results.sort((a, b) => b.final_score - a.final_score);

    // Return top results
    return results.slice(0, limit);
  } catch (error) {
    console.error("[SemanticSearch] Error during search:", error);
    return [];
  } finally {
    client.release();
  }
}

/**
 * Simple semantic search using just use_case query
 * This is the most common search pattern
 */
export async function searchByUseCase(
  useCase: string,
  options: SemanticSearchOptions = {}
): Promise<SemanticSearchResult[]> {
  return semanticSearchStructured({ use_case: useCase }, options);
}

/**
 * Check if semantic search tables are available
 */
export async function isSemanticSearchAvailable(): Promise<boolean> {
  const client = await getClient();
  if (!client) return false;

  try {
    const result = await client.query(
      `SELECT COUNT(*)::int AS count FROM public.kb_embeddings LIMIT 1`
    );
    return (result.rows[0]?.count ?? 0) > 0;
  } catch (error) {
    console.error("[SemanticSearch] Error checking availability:", error);
    return false;
  } finally {
    client.release();
  }
}

/**
 * Get common ambiguity triggers from a set of results
 * Used to determine if clarifying questions are needed
 */
export function extractCommonAmbiguityTriggers(
  results: SemanticSearchResult[],
  threshold: number = 0.5
): string[] {
  if (results.length === 0) return [];

  // Count occurrences of each trigger
  const triggerCounts: Record<string, number> = {};
  for (const result of results) {
    for (const trigger of result.ambiguity_triggers) {
      triggerCounts[trigger] = (triggerCounts[trigger] || 0) + 1;
    }
  }

  // Return triggers that appear in more than threshold % of results
  const minCount = Math.ceil(results.length * threshold);
  return Object.entries(triggerCounts)
    .filter(([, count]) => count >= minCount)
    .map(([trigger]) => trigger);
}

/**
 * Map price tier to budget range
 */
export function priceTierToBudgetRange(tier: string): { min?: number; max?: number } {
  const ranges: Record<string, { min?: number; max?: number }> = {
    budget: { max: 5000 },
    low: { min: 5000, max: 10000 },
    mid: { min: 10000, max: 20000 },
    high: { min: 20000, max: 50000 },
    premium: { min: 50000 },
  };
  return ranges[tier.toLowerCase()] || {};
}

/**
 * Map budget to price tier
 */
export function budgetToPriceTier(budget: { min?: number; max?: number }): string | null {
  if (budget.max && budget.max <= 5000) return "budget";
  if (budget.max && budget.max <= 10000) return "low";
  if (budget.max && budget.max <= 20000) return "mid";
  if (budget.max && budget.max <= 50000) return "high";
  if (budget.min && budget.min >= 50000) return "premium";
  if (budget.min && budget.min >= 20000) return "high";
  if (budget.min && budget.min >= 10000) return "mid";
  return null;
}




